prepress.py

```diff
@@ -255,17 +255,98 @@ def replace_dashes(article: Article) -> Article:
         text_tag.replace_with(new_tag)
     return article
 
+# Quotes only really start with varieties of left brackets
+LEFT_PUNCTUATION = { '(', '{', '<' }
+# But they can be ended with any of the standard sentence enders
+RIGHT_PUNCTUATION = { ')', '}', '>', '.', ',', '?', '!' }
+LEFT_DOUBLE_QUOTE = '“'
+RIGHT_DOUBLE_QUOTE = '”'
+STRAIGHT_DOUBLE_QUOTE = '"'
+
+def replace_double_quote(before: str, after: str):
+    # double quotes can occur at the beginning/end of strings
+    if before is None:
+        return LEFT_DOUBLE_QUOTE
+    if after is None:
+        return RIGHT_DOUBLE_QUOTE
+
+    # double quotes can occur before/after spaces
+    if before.isspace():
+        return LEFT_DOUBLE_QUOTE
+    # ditto for the right
+    if after.isspace():
+        return RIGHT_DOUBLE_QUOTE
+
+    # double quotes can occur before/after punctuation
+    if before in LEFT_PUNCTUATION:
+        return LEFT_DOUBLE_QUOTE
+    if after in RIGHT_PUNCTUATION:
+        return RIGHT_DOUBLE_QUOTE
+
+
+    return STRAIGHT_DOUBLE_QUOTE
+
+LEFT_SINGLE_QUOTE = '‘'
+RIGHT_SINGLE_QUOTE = '’'
+STRAIGHT_SINGLE_QUOTE = '\''
+
+def replace_single_quote(before: str, after: str):
+    # the rules for single quotes are the same as double quotes
+    double_quote = replace_double_quote(before, after)
```

Comment:
We should split the functionality for determining what quote to return into a separate function that returns an enum or sentinel value, e.g.

```python
class QuoteRules:
    VERBATIM = 0
    LEFT = 1
    RIGHT = 2


def should_replace_quote(before: str, after: str):
    # ...logic...


def replace_double_quote(...):
    replace_rule = should_replace_quote(...)
    if replace_rule == QuoteRules.LEFT:
        ...
    # and so on
```